src/main/java/com/powsybl/openloadflow/util/PropagatedContingency.java

```diff
@@ -46,55 +49,88 @@ public int getIndex() {
+                    contingency.getId(), coupler.get().getId());
             }
+        }
+        return propagatedContingencies;
+    }
 
-            for (Switch sw : switchesToOpen) {
-                propagatedContingency.getBranchIdsToOpen().add(sw.getId());
-                allSwitchesToOpen.add(sw);
-            }
+    public static List<PropagatedContingency> createListForSecurityAnalysis(Network network, List<Contingency> contingencies, Set<Switch> allSwitchesToOpen) {
+        List<PropagatedContingency> propagatedContingencies = new ArrayList<>();
+        for (int index = 0; index < contingencies.size(); index++) {
+            Contingency contingency = contingencies.get(index);
+            PropagatedContingency propagatedContingency = PropagatedContingency.create(network, contingency, index);
+            propagatedContingencies.add(propagatedContingency);
+            allSwitchesToOpen.addAll(propagatedContingency.switchesToOpen);
+        }
+        return propagatedContingencies;
+    }
 
-            for (Terminal terminal : terminalsToDisconnect) {
-                if (terminal.getConnectable() instanceof Branch) {
-                    propagatedContingency.getBranchIdsToOpen().add(terminal.getConnectable().getId());
-                }
+    private static PropagatedContingency create(Network network, Contingency contingency, int index) {
+        Set<Switch> switchesToOpen = new HashSet<>();
+        Set<Terminal> terminalsToDisconnect =  new HashSet<>();
+        Set<String> branchIdsToOpen = new HashSet<>();
+        Set<String> hvdcIdsToOpen = new HashSet<>();
+        for (ContingencyElement element : contingency.getElements()) {
+            switch (element.getType()) {
+                case BRANCH:
+                case LINE:
+                case TWO_WINDINGS_TRANSFORMER:
+                    // branch check is done inside branch tripping
+                    new BranchTripping(element.getId(), null)
+                        .traverse(network, null, switchesToOpen, terminalsToDisconnect);
+                    branchIdsToOpen.add(element.getId());
+                    break;
+                case HVDC_LINE:
+                    HvdcLine hvdcLine = network.getHvdcLine(element.getId());
+                    if (hvdcLine == null) {
+                        throw new PowsyblException("HVDC line '" + element.getId() + "' not found");
+                    }
+                    hvdcIdsToOpen.add(element.getId());
+                    break;
+                default:
+                    //TODO: support all kinds of contingencies
+                    throw new UnsupportedOperationException("TODO");
             }
-
         }
-        return propagatedContingencies;
+
+        return new PropagatedContingency(contingency, index, branchIdsToOpen, hvdcIdsToOpen, switchesToOpen, terminalsToDisconnect);
+    }
+
+    private static boolean isCoupler(Switch s) {
```

Comment:
`hasBusBarSectionCoulper` could be better here, what do you think?

Reply:
for me coupler already means busbar section coupler